astrobee/behaviors/inspection/include/inspection/camera_projection.h

```diff
@@ -0,0 +1,126 @@
+/* Copyright (c) 2021, United States Government, as represented by the
+ * Administrator of the National Aeronautics and Space Administration.
+ *
+ * All rights reserved.
+ *
+ * The "ISAAC - Integrated System for Autonomous and Adaptive Caretaking
+ * platform" software is licensed under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+#ifndef INSPECTION_CAMERA_PROJECTION_H_
+#define INSPECTION_CAMERA_PROJECTION_H_
+
+// Standard ROS includes
+#include <nodelet/nodelet.h>
+#include <pluginlib/class_list_macros.h>
+#include <ros/ros.h>
+
+// TF2 support
+#include <tf2_ros/transform_listener.h>
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+
+// Shared project includes
+#include <ff_common/ff_names.h>
+#include <msg_conversions/msg_conversions.h>
+#include <config_reader/config_reader.h>
+#include <isaac_util/isaac_names.h>
+
+// Point Cloud
+#include <sensor_msgs/PointCloud2.h>
+#include <sensor_msgs/point_cloud_conversion.h>
+
+// Software messages
+#include <visualization_msgs/MarkerArray.h>
+#include <geometry_msgs/PoseArray.h>
+
+// Eigen for math
+#include <Eigen/Dense>
+
+// C++ headers
+#include <vector>
+#include <string>
+#include <map>
+
+/**
+ * \ingroup beh
+ */
+namespace inspection {
+
+/*
+  This class provides camera functionality that allows us
+  to project the 3D point into the camera frame and the
+  other way around. It automatically reads the camera
+  parameters from the config files based on the camera
+  name, such that no setup is necessary.
+*/
+class CameraView {
+ public:
+  // Constructor
+  explicit CameraView(const std::string cam_name, const float f = 2.0, const float n = 0.19,
+                      const geometry_msgs::Transform::ConstPtr cam_transform = NULL);
+
+  Eigen::Matrix4d GetProjectionMatrix();
+  double GetHFOV();
+  double GetVFOV();
+
+  double GetH();
+  double GetW();
+
+  bool SetH(const double H);
+  bool SetW(const double W);
+
+  bool BuildViewMatrix(const geometry_msgs::Pose robot_pose, Eigen::Matrix4d &V);
+  bool GetVectorFromCamXY(const geometry_msgs::Pose robot_pose, const int x, const int y,
```

Comment:
the distortion matrix is part of the class in the var P_ and is used in the calculation

Reply:
Yeah I'm tempted to delay this and merge the solution of this with https://github.com/nasa/isaac/issues/71.
My overall plan is to make `CameraView` inherit from `CameraModel` and use all the common functionality. The current issue is that `CameraModel` does not consider some parameters in the projection matrix like the `CameraView` (the `s`, `cx` and `cy`), only focal vector (any change to `CameraModel` needs to be thoroughly tested and evaluate the impact across fsw. Moreover the `IsInFov()` has a comment which is worth looking at:

````
// TODO(oalexan1): This looks buggy. Because ImageCoordinates()
// returns an undistorted pixel, it must compare to GetUndistortedHalfSize().
````

Reply:
That seems reasonable, but in that case we should add a TODO comment that mentions issue #71. I also added a comment there about a tweak to the approach that should address Oleg's concern.